File: 周赛/【303周赛】2353-设计食物评分系统.py
from sortedcontainers import SortedList
class FoodRatings:

    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        # 存储由food到cuisines和rating的映射，是一对一关系
        self.f2cr = dict() 
        # 存储由cuisines到food和rating的映射，是一对多关系
        self.c2rf = dict() # 因为答案会取先按rating排名后按食物名称字典序排名，因此该dict在存储时要先存rating，后存food
        for food, cuisine, rating in zip(foods, cuisines, ratings):
            self.f2cr[food] = [cuisine, rating]
            if cuisine not in self.c2rf:
                self.c2rf[cuisine] = SortedList()
            self.c2rf[cuisine].add([-rating, food]) # rating取负存储可使大值排在前面

    def changeRating(self, food: str, newRating: int) -> None:
        preRate = self.f2cr[food][1]
        cuisine = self.f2cr[food][0]
        if preRate == newRating: return
        else:
            # 新分数与历史分数不等的情况下，需要删除cuisines映射中的原数据，两次更新将在if外再执行
            self.c2rf[cuisine].remove([-preRate, food])
            self.f2cr[food][1] = newRating
            self.c2rf[cuisine].add([-newRating, food])


    def highestRated(self, cuisine: str) -> str:
        return self.c2rf[cuisine][0][1]


# Your FoodRatings object will be instantiated and called as such:
# obj = FoodRatings(foods, cuisines, ratings)
# obj.changeRating(food,newRating)
# param_2 = obj.highestRated(cuisine)

File: 周赛/【303周赛】2353-设计食物评分系统.cpp
class FoodRatings {
    typedef pair<int, string> pis;
    map<string, pis> f2rc;
    map<string, set<pis>> c2rf;
public:
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        for(int i=0;i<foods.size();i++)
        {
            f2rc[foods[i]]=pis(ratings[i],cuisines[i]);
            c2rf[cuisines[i]].insert(pis(-ratings[i],foods[i]));
        }
    }
    
    void changeRating(string food, int newRating) {
        pis &p=f2rc[food];
        c2rf[p.second].erase(pis(-p.first,food));
        p.first=newRating;
        c2rf[p.second].insert(pis(-newRating,food));
    }
    
    string highestRated(string cuisine) {
        return c2rf[cuisine].begin()->second;
    }
};

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings* obj = new FoodRatings(foods, cuisines, ratings);
 * obj->changeRating(food,newRating);
 * string param_2 = obj->highestRated(cuisine);
 */
